src/config.rs

```diff
@@ -39,33 +51,194 @@ pub struct Config {
+        let rule_arg = self.rule.as_ref().unwrap();
+        if self
+            .all_rules
+            .iter()
+            .any(|rule| &rule.meta().name == rule_arg)
+        {
+            panic!("The '{rule_arg}' rule is configured as inactive");
+        } else {
+            panic!("Unknown rule: '{rule_arg}'");
+        }
     }
 
     pub fn get_instantiated_rules(&self) -> Vec<InstantiatedRule> {
-        let instantiated_rules = self
-            .rules()
+        let active_rules = self.get_active_rules();
+        if active_rules.is_empty() {
+            panic!("No configured active rules");
+        }
+        let active_rules = self.filter_based_on_rule_argument(active_rules);
+        let instantiated_rules = active_rules
             .into_iter()
             .map(|rule| InstantiatedRule::new(rule.clone(), self))
-            .filter(|rule| match self.rule.as_ref() {
-                Some(rule_arg) => &rule.meta.name == rule_arg,
-                None => true,
-            })
             .collect::<Vec<_>>();
         if instantiated_rules.is_empty() {
             panic!("Invalid rule name: {:?}", self.rule.as_ref().unwrap());
         }
         instantiated_rules
     }
 }
+
+impl ConfigBuilder {
+    pub fn default_rule_configurations(&mut self) -> &mut Self {
+        self.rule_configurations = Some(
+            self.all_rules
+                .as_ref()
+                .expect("must call .all_rules() before calling .default_rule_configurations()")
+                .into_iter()
+                .map(RuleConfiguration::default_for_rule)
+                .collect(),
+        );
+        self
+    }
+}
+
+#[derive(Clone)]
+pub struct ParsedConfigFile {
+    pub path: PathBuf,
+    pub content: ParsedConfigFileContent,
+}
+
+#[derive(Clone, Deserialize)]
+pub struct ParsedConfigFileContent {
+    #[serde(rename = "rules")]
+    rules_by_name: HashMap<String, RuleConfigurationValue>,
+}
+
+impl ParsedConfigFileContent {
+    pub fn rules(&self) -> impl Iterator<Item = RuleConfiguration> + '_ {
+        self.rules_by_name
+            .iter()
+            .map(|(rule_name, rule_config)| rule_config.to_rule_configuration(rule_name))
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub enum ErrorLevel {
+    Error,
+    Off,
+}
+
+#[derive(Clone, Deserialize)]
+pub struct RuleConfigurationValue {
+    pub level: ErrorLevel,
+}
+
+impl RuleConfigurationValue {
+    pub fn to_rule_configuration(&self, rule_name: impl Into<String>) -> RuleConfiguration {
+        let rule_name = rule_name.into();
+        RuleConfiguration {
+            name: rule_name,
+            level: self.level,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct RuleConfiguration {
+    pub name: String,
+    pub level: ErrorLevel,
+}
+
+impl RuleConfiguration {
+    pub fn default_for_rule(rule: &Arc<dyn Rule>) -> Self {
+        Self {
+            name: rule.meta().name,
```

Comment:
These repeated calls to `.meta()` are bothering me because they re-allocate and conceptually the `meta` is just a static "singleton" per `Rule`

So at the moment I'm a little tempted to eg wrap (the return value of `Rule::meta`, and then `Rule` implementors would presumably have an internal `meta: Arc<RuleMeta>` field that they just `.clone()`) in an `Arc`?